shadowsocks-csharp/Controller/Service/TCPRelay.cs

```diff
@@ -37,9 +45,24 @@ public override bool Handle(byte[] firstPacket, int length, Socket socket, objec
         {
             if (socket.ProtocolType != ProtocolType.Tcp
                 || (length < 2 || firstPacket[0] != 5))
+            {
                 return false;
+            }
+
             socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.NoDelay, true);
-            TCPHandler handler = new TCPHandler(_controller, _config, this, socket);
+            TCPHandler handler = new TCPHandler(_controller, _config, socket);
+
+            handler.OnConnected += OnConnected;
+            handler.OnInbound += OnInbound;
+            handler.OnOutbound += OnOutbound;
+            handler.OnFailed += OnFailed;
+            handler.OnClosed += (h, arg) =>
+            {
+                lock (Handlers)
+                {
+                    Handlers.Remove(handler);

```

Comment:
May need to remove all registered events (`OnConnected`, `OnInbound`, `OnOutbound`, `OnFailed`, `OnClosed`)

Reply:
I don't think so, re-trigger following event after `OnClosed` won't cause problem.